client/analytics/settings/index.js

```diff
@@ -206,22 +206,62 @@ class Settings extends Component {
 }
 
 export default compose(
-	withSelect( select => {
-		const { getSettings, getSettingsError, isGetSettingsRequesting } = select( 'wc-api' );
+	createHigherOrderComponent(
+		WrappedComponent => props => {
+			const settings = useRef( getSetting( 'preloadSettings', { wc_admin: {} } ) );
+			useSelect( ( select, registry ) => {
+				if ( ! settings.current ) {
+					return;
+				}
 
-		const settings = getSettings( 'wc_admin' );
-		const isError = Boolean( getSettingsError( 'wc_admin' ) );
-		const isRequesting = isGetSettingsRequesting( 'wc_admin' );
+				const { isResolving, hasFinishedResolution } = select( SETTINGS_STORE_NAME );
+				const { startResolution, finishResolution, hydrateSettings } = registry.dispatch(
+					SETTINGS_STORE_NAME
+				);
 
-		return { getSettings, isError, isRequesting, settings };
-	} ),
+				if (
+					! isResolving( 'getSettings', [ 'wc_admin' ] ) &&
+					! hasFinishedResolution( 'getSettings', [ 'wc_admin' ] )
+				) {
+					startResolution( 'getSettings', [ 'wc_admin' ] );
+					hydrateSettings( settings.current );
+					finishResolution( 'getSettings', [ 'wc_admin' ] );
+				}
+			}, [] );
+			return <WrappedComponent { ...props } />;
+		},
+		'withHydration'
```

Comment:
@nerrad is the idea here that `withHydration` will be an exportable HOC? So any component looking to interact with settings could use this?

Reply:
It'd be something primarily wc-admin would use to hydrate it's stores from the server.  So essentially you'd use the hoc to wrap the App entry point so that store hydration is automatically taken care of (and anything else using the store in that view would be interacting with a already hydrated store assuming they are listing wc-admin stores as a dependency).

It feasibly _could_ be abstracted so that it could be exported as a utility other client code could use to hydrate store values but I'd start simple and keep it for internal wc-admin use initially (and thus just hydrating client).  You can see the blocks project usage:  [HOC here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/hocs/with-rest-api-hydration.js) and implementation [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/assets/js/blocks/products/all-products/frontend.js)

Reply:
Also, noting this would be where we interchange `useSelect` with `useSelectWithRefresh`.

I think we can leave this aspect up to the code that handles https://github.com/woocommerce/woocommerce-admin/issues/3405 because this will be likely to change as a result.

Reply:
`preloadSettings` is an onboarding thing, which seems to be gathered from data written to `wcSettings.preloadSettings`. So, we can definitely remove from here and update the OBW to also use this new data package. I'm pretty sure we can remove the `getSettings` resolver because it doesn't appear to be used, on master as well.

In fact `store.getSettings( 'wc_admin' )` can be hard coded to an empty object and the page works normally. This has me thinking 🤔, whats the point of HOC `withSettings` anyways? All settings are gathered from `@woocommerce/wc-admin-settings` instead of the `settings` prop. 

The `persistAllSettings` action is useful as it can make a request to update settings on on the server as well as the object on the window via `setSetting`. 

@nerrad how do you feel about moving all the logic currently in `client/settings` to this store? All areas of the app that need to interact with settings can do so via the `withSettings` HOC. 

I think a good course of action will be to remove `withHydration` from this PR, re-add fresh-data settings api (when I removed it, OBW broke), and ready this for final review while making the above changes as part of https://github.com/woocommerce/woocommerce-admin/issues/3405

Reply:
> I'm pretty sure we can remove the getSettings resolver because it doesn't appear to be used, on master as well.

You can remove this if you want, but it's handy having it for if this store get's used by third party extensions and they don't pre-hydrate settings for their view (or it's an environment where pre-hydrating isn't as easily done).

> This has me thinking 🤔, whats the point of HOC withSettings anyways? All settings are gathered from @woocommerce/wc-admin-settings instead of the settings prop.

This is why my comments about the number of settings states and #3405 exists.  In my opinion, for the settings managed by this store `@woocommerce/wc-admin-settings` should ONLY be used for hydrating the initial store state and then from that point on the store should serve as the authority.  When you use the store to persist setting changes to the server, the store state update will automatically signal to any components subscribed to the store (via `withSelect` or `useSelect` mapped to the store selectors) to update.  `@woocommerce/wc-admin-settings` is a poor mechanism for communicating these changes in state.

> The persistAllSettings action is useful as it can make a request to update settings on on the server as well as the object on the window via setSetting.

If you're talking about updating the window object from `persistAllSettings` I don't think that's a good idea.  You'd be coupling window globals to the data store.   In fact, I don't think this window object for settings _should_ be updated at all.  Related to previous conversations I've had about this, I think the only purpose of for the window global is to hydrate the initial settings state from the server (to avoid additional requests).

> @nerrad how do you feel about moving all the logic currently in client/settings to this store? All areas of the app that need to interact with settings can do so via the withSettings HOC.

This seems to be in line with what I'm advocating for. As long as you're not talking about also moving interactions with the window global to the store too.

> I think a good course of action will be to remove withHydration from this PR, re-add fresh-data settings api (when I removed it, OBW broke), and ready this for final review while making the above changes as part of #3405

- re `withHydration` removal... you'll have to make the call on that.  I'm guessing though that preLoad settings would still have a value during a specific state of onboarding?  I was never able to actually see data in that object coming from the server but maybe that was just an artifact of how I was testing.  
- re-add fresh-data settings api:  yeah, should only remove it once you've confirmed anything using it has been decoupled from it and is using the new store.
- re: making above changes part of #3405.  👍

Reply:
Thanks for the feedback @nerrad. 

>In my opinion, for the settings managed by this store @woocommerce/wc-admin-settings should ONLY be used for hydrating the initial store state and then from that point on the store should serve as the authority

Absolutely, that makes sense to me now.

>re withHydration removal... you'll have to make the call on that. I'm guessing though that preLoad settings would still have a value during a specific state of onboarding? I was never able to actually see data in that object coming from the server but maybe that was just an artifact of how I was testing

`preloadSettings` is treated just like any other subset of settings. Its loaded onto the settings object:

https://github.com/woocommerce/woocommerce-admin/blob/0860b3b2b3bef9d91906e02cd6ac6fe01a1f1dd2/src/Loader.php#L638

Then retrieved from `wcSettings` like other properties. I'm going to remove it from the settings page, and if I'm thinking about it incorrectly, circle back to `withHydration`.